networkit/cpp/centrality/GroupHarmonicCloseness.cpp

```diff
@@ -0,0 +1,695 @@
+/*
+ * GroupHarmonicCloseness.cpp
+ *
+ * Created on: 15.12.2020
+ *     Author: Eugenio Angriman <angrimae@hu-berlin.de>
+ */
+
+// networkit-format
+
+#include <atomic>
+#include <cassert>
+#include <cmath>
+#include <limits>
+#include <omp.h>
+#include <queue>
+#include <stdexcept>
+
+#include <networkit/centrality/GroupHarmonicCloseness.hpp>
+#include <networkit/components/ConnectedComponents.hpp>
+#include <networkit/graph/BFS.hpp>
+#include <networkit/graph/Dijkstra.hpp>
+#include <networkit/reachability/ReachableNodes.hpp>
+#include <networkit/structures/Partition.hpp>
+
+namespace NetworKit {
+
+namespace {
+
+struct SSSPResult {
+    bool complete;
+    double score;
+    SSSPResult(bool complete, double score) : complete(complete), score(score) {}
+};
+
+template <class WeightType>
+class GroupHarmonicClosenessImpl final
+    : public GroupHarmonicCloseness::GroupHarmonicClosenessInterface {
+
+    static constexpr WeightType infDist = std::numeric_limits<WeightType>::max();
+
+public:
+    GroupHarmonicClosenessImpl(const Graph &G, count k = 1);
+
+    ~GroupHarmonicClosenessImpl() override;
+
+    void run() override;
+
+    static double scoreOfGroup(const Graph &graph, const std::vector<node> &inputGroup);
+
+private:
+    const Graph *G;
+    const count k;
+    WeightType minEdgeWeight = infDist;
+
+    std::unique_ptr<Partition> component;
+
+    std::vector<WeightType> distFromGroup, curBestDist;
+    std::vector<std::vector<WeightType>> curDistGlobal;
+    std::vector<count> reachableNodesInComponent, reachableNodesUB;
+    std::vector<node> pred, nearerNodes;
+    std::vector<std::vector<node>> curNearerNodesGlobal;
+    std::vector<double> margGain;
+    std::vector<std::vector<bool>> visitedGlobal;
+
+    template <class Type>
+    struct Greater {
+        Greater(const std::vector<Type> &margGain) : margGain(margGain) {}
+        bool operator()(node x, node y) const noexcept { return margGain[x] > margGain[y]; }
+
+    private:
+        const std::vector<Type> &margGain;
+    };
+
+    template <class Type>
+    struct Less {
+        Less(const std::vector<Type> &dist) : dist(dist) {}
+        bool operator()(node x, node y) const noexcept { return dist[x] < dist[y]; }
+
+    private:
+        const std::vector<Type> &dist;
+    };
```

Comment:
Should these templates be in `Aux` such that multiple algos can use them? (See other PR.)